Documentation revision (#25)
PerfectFit-project/virtual-coach-main: docs/design.md

### Expiration of data
Some data might only be needed for a certain time period, especially sensor data. 
A specific worker should remove data if it is expired (data should have an expiry-date field in db).

---

## Architecture sketch
<img src = "img/design.png" width = "1000" title="Design">

See it also on [whimsical](https://whimsical.com/perfectfit-UtvRnxdP8P79humXTnjb9J)

---

## Functionalities
An overview of the functionalities built into PerfectFit is shown here below. These repositories implement the chatbot conversation functionality.

<img src = "img/usecase_diagram.png" width = "500" title="Use case">

The chatbot uses a set of components that allow the conversation to flow in the intended way, consuming and saving data from different sources (i.e., Niceday server and local DB). A representation of the interactions among the components in the conversation flow is shown here below:

> <img src = "img/swimlane.png" width = "1000" title="Swimlane">

Reviewer: To me this figure is already very helpful - but of course I do not know if something is missing at this point.

When a new message is received, the broker composes a message according to the rasa server interface specifications and sends it to the Rasa server though an HTTP POST request. The response to the HTTP request contains the list of messages produced by the Rasa Bot. Upon the response reception, the broker sends each single message to the Niceday server using the sendTextMessage(id, text) function of the Chat object of [goalie-js](https://github.com/senseobservationsystems/goalie-js).

> <img src = "img/sequence.png" width = "500" title="Sequence diagram">

Reviewer: Here I was wondering why there are two requests to rasa actions.

> <img src = "img/sequence.png" width = "500" title="Sequence diagram">

Author: @nelealbers the two requests were intended as an example of multiple actions executed to respond to one message. I realize that this is not clear, and I will specify this in the text.

### Data exchange with Niceday Server

To retrieve data from the Niceday Server, the [Niceday API](https://github.com/PerfectFit-project/niceday-components/tree/main/niceday-api) has to be used. The  API is used by the Rasa actions, using the [NicedayClient](https://github.com/PerfectFit-project/niceday_client) Python interface. 

<img src = "img/sequence_data.png" width = "500" title="Sequence diagram data">

### Data exchange with local Database

User’s data and output of the conversations can be stored to and retrieved from the local DB, represented here below. The interactions with the DB are performed by the rasa actions, using the [virtual-coach-db](https://github.com/PerfectFit-project/virtual-coach-db) package.

> <img src = "img/db.png" width = "300" title="Sequence diagram data">

Reviewer: I also like the diagram (in addition to what is in models.py). But indeed important to tell people to maintain this.

> <img src = "img/db.png" width = "300" title="Sequence diagram data">

Author: I personally prefer to have a visual representation of the schema. It will help the understanding of the DB especially if we are going to extend it. 
I agree that the file has to referenced also to allow the maintenance and that it makes more sense to have it in the virtual-coach-db repo.

---

## Interfaces specifications

The different components exchange information one with the other using HTTP requests. The requests specifications are reported here below.

> N.B. The URLs  reported are assuming that the component is running using the Docker image generated using the [docker-compose file](https://github.com/PerfectFit-project/virtual-coach-main/blob/main/docker-compose.yml). If the components are running in isolation, or outside the Docker network, the URL has to be changed accordingly. 

Reviewer: Would help to provide examples of how the URLs have to be changed accordingly.

> ###Rasa server

Reviewer 2: ```suggestion
### Rasa server
```

URL: http://rasa_server:5005/webhooks/rest/webhook

Request type: POST

Body:

{
    "sender":"sender_id",
    "message":"text_message"
}

> 

Reviewer 2: Please put this in a 
``` 
codeblock
```

Output:

[
    {
        "recipient_id": "id",
        "text": "text_message"
    },
    {
        "recipient_id": "id",
        "text": " text_message”
    }

> ]

Reviewer 2: Also please put this in a codeblock, so it is rendered nicely. See https://github.com/PerfectFit-project/virtual-coach-main/blob/documentation-revision/docs/design.md#interfaces-specifications

> ###Rasa actions

Reviewer 2: ```suggestion
### Rasa actions
```

URL: http://rasa_actions:5005/webhook

[Specifications](https://rasa.com/docs/rasa/custom-actions/)

> ###Niceday API

Reviewer 2: ```suggestion
### Niceday API
```

URL: http://niceday_api:8080/

> Specifications for the API usage can be found in the Swagger UI interface, as shown [here](https://github.com/PerfectFit-project/niceday-components/tree/main/niceday-api/README.md).

Reviewer 2: ```suggestion
Specifications for the API usage can be found in the Swagger UI interface, as shown 
[here](https://github.com/PerfectFit-project/niceday-components/blob/main/niceday-api/README.md#running-the-server).
```

> ###Broker

Reviewer 2: ```suggestion
### Broker
```
